appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockDataFile.java

```diff
@@ -0,0 +1,189 @@
+    super(editor, type, iconImage);
+
+    dataFileChangeListeners = new HashSet<DataFileChangeListener>();
+  }
+
+  /**
+   * Changes the Source File property of the MockDataFile.
+   * Upon change, the new file is imported and parsed, the
+   * contents are stored, and DataFileChangeListeners are
+   * updated.
+   *
+   * @param fileSource  name of the new Source File
+   */
+  private void setSourceFileProperty(String fileSource) {
+    // Update the source file property & reset the
+    // columnNames and rows property
+    this.sourceFile = fileSource;
+    columnNames = new ArrayList<String>();
+
+    // Update DataFileChangeListeners to notify that
+    // the columns list is (at least temporarily) empty
+    updateDataFileChangeListeners();
+
+    // Check that the SourceFile property is a valid file
+    if (fileSource == null || fileSource.equals("")) {
+      return;
+    }
+
+    // Read the media file
+    long projectId = editor.getProjectId();
+
+    Ode.getInstance().getProjectService().loadDataFile(projectId, "assets/" + fileSource,
+        new AsyncCallback<List<List<String>>>() {
+      @Override
+      public void onFailure(Throwable caught) {
+        ErrorReporter.reportError(caught.getMessage());
+      }
+
+      @Override
+      public void onSuccess(List<List<String>> result) {
+        // Update columns property
+        columns = result;
+
+        if (result.isEmpty()) {
+          ErrorReporter.reportError(MESSAGES.emptyFileError());
+          return;
+        }
+
+        // Populate columnNames List (it was empty before)
+        for (int i = 0; i < columns.size(); ++i) {
+          List<String> column = columns.get(i);
+
+          // Add first element of the column if the column is not empty
+          if (!column.isEmpty()) {
+            columnNames.add(column.get(0));
+          }
+        }
+
+        // Notify DataFileChangeListeners of the changes
+        updateDataFileChangeListeners();
+
+        // Hide the info message shown after setting the Source File property
+        ErrorReporter.hide();
+      }
+    });
+
+    // Show message to indicate parsing of the files
+    // (since this is an asynchronous operation)
+    ErrorReporter.reportInfo(MESSAGES.dsataFileParsingMessage(sourceFile, this.getName()));
+  }
+
+  /**
+   * Get the Column Names (first row) of the MockDataFile's parsed content
+   *
+   * @return  column names of the Data File (list of Strings)
+   */
+  public List<String> getColumnNames() {
+    return columnNames;
+  }
+
+  /**
+   * Returns a List of the specified columns.
+   *
+   * If a column is not found, it is substituted by
+   * an empty List.
+   *
+   * @param columns  List of columns to get
+   * @return  List of columns (a column is a List of Strings)
+   */
+  public List<List<String>> getColumns(List<String> columns) {
+    ArrayList<List<String>> dataColumns = new ArrayList<List<String>>();
```

Comment:
@ewpatton I don't think this is worth caching, since the columns List already caches all the columns, and this method is used to retrieve a subset of columns.